discord/gateway.py

```diff
@@ -727,7 +727,7 @@ def __init__(self, *args, **kwargs):
         log.debug('detected ip: %s port: %s', state.ip, state.port)
 
         # there *should* always be at least one supported mode (xsalsa20_poly1305)
-        modes = [mode for mode in data['modes'] if mode in self._connection.supported_modes]
+        modes = list(set(self._connection.supported_modes).intersection(data['modes']))
```

Comment:
I don't see why this change was necessary.

Reply:
I'm in the process of reverting it. In testing `xsalsa20_poly1305` was still being chosen over preferable modes. However more closely examining results from this change shows this also does not fix that consistently.